### ISSUE TITLE
Bury order in v3 queue build is incorrect, leading to "bug: card modified" messages

### ISSUE THREAD

**Issue by dae** [ankitects/anki#2352]
I suspect the issue mentioned on https://forums.ankiweb.net/t/card-modified-without-updating-queue/26052 was caused by the following configuration:

- bury reviews: yes
- bury new: no
- mix reviews and new cards

In gathering.rs, we gather the review cards first, and check if they should be buried at that point. We then gather the new cards. Only after gathering do we sort the cards in the order they'll appear in to the user. If a new sibling happens to be shown first, that will result in the review card being buried, but it's not removed from the queue, as the code expects it to have never been included.

To solve this, we'll presumably need to shift the bury check to apply as we collect the merge_new() iterator into a vec, so that we bury in exactly the same order as cards will appear. I think we may be ok with keeping the intraday learning code as is, since it comes first, and intraday cards are not subject to burying themselves.

@RumovZ interested in tackling this one?

**Comment by RumovZ** [comment]
Yeah, sure!

**Comment by RumovZ** [comment]
To prove your hypothesis, I created a deck with one review and one new sibling, set _New/review order_ to _Show before reviews_, checked _Bury review siblings_ and unchecked _Bury new siblings_. This does indeed lead to an error. Also, if I check _Bury new siblings_, the new sibling gets buried, but, arguably, it should be the review sibling. (A minor issue, but will be relevant later.)

Now, as for your suggestion, don't we have to bury while gathering, as otherwise we don't know how many cards to gather?

Intuitively, I'd say we'd have to gather in the order of reviews. So if new cards are shown before reviews, we should gather new cards first, then bury review cards accordingly. However, that poses a problem with limits: We could end up with too many new cards after the new limit has been capped to the review limit. And we can't just discard some of the new cards, because we have already buried their review siblings.
I think we have a deadlock here, where both the gathering of review cards and the gathering of new cards depend on each other.

Solutions I can think of right now:
1. Prohibit certain combinations of settings.
    - Disadvantage 1: Requires the various frontends to enforce it.
    - Disadvantage 2: Users might (rightfully) expect these combinations to work.
2. Repeat queue building steps in a loop.
    - Disadvantage 1: Adds complexity.
    - Disadvantage 2: Less efficient in some cases.


Solution 2 in the context of this issue:
1. Gather new cards.
2. Gather review cards.
3. If too many new cards have been gathered, reduce their amount accordingly, reset bury state and go to 2.

**Comment by dae** [comment]
Thanks for looking into this Rumo.

Unfortunate that there's no easy fix for this - the queue building code already is somewhat complex, and multiple passes is going to make the code harder to reason about moving forward, so suggestion 1 seems more appealing at this stage.

> Also, if I check Bury new siblings, the new sibling gets buried, but, arguably, it should be the review sibling.

I can definitely see how the user might expect that, given the selected order. But in terms of what's best for memory/keeping up with reviews, I think a good argument could be made for prioritizing reviews over new cards, and we already do that when it comes to determining the daily new counts.

Thinking about it more, perhaps the same issue exists for interday learning cards as well? If burying of interday cards is on but reviews is off, and the review card is shown first, that's going to presumably cause the same problem. 

Perhaps the solution here is to ensure that options are set in the order we gather? Eg enabling review card burying implies new card burying, and enabling interday burying implies review card burying + new card burying?

> Disadvantage 1: Requires the various frontends to enforce it.

With the recently-added `ensure_values_valid()`, we can apply corrections to deck configs as they're loaded, so it would be easy enough to enable some of the booleans if others were set.

In terms of updating the UI, the legacy deck options screen doesn't have any v3-related options, so I'm not concerned about that. Two possible approaches I can think of for the UI:

a) Automatically enable some of the options when others are selected, and display a message when an option is disabled that must be enabled.
b) Change the three switches into a dropdown (new / new+review / new+review+interday). To avoid backwards-compatibility issues, probably easiest to keep the separate booleans on disk and have the enum only used in the UI.

> Disadvantage 2: Users might (rightfully) expect these combinations to work.

Yep, that's a fair point. It is more restrictive than the v2 scheduler, where reviews and new cards could be configured independently, but we didn't have to worry about calculating bury counts up front back then.

We could potentially relax the restrictions if there turned out to be a lot of unhappy users, at least if certain sort orders were selected. That might be an argument for approach a) instead of b), since the latter would not allow for independent control.

What do you think? I don't have strong opinions at this point, and am open to other ideas/pushback.

**Comment by RumovZ** [comment]
After thinking long and hard about this, I think you are right.

Regardless of the order of appearance, it's generally more reasonable to bury new cards and avoid burying due cards where possible. Even if there is a use case for the reverse, we'd need yet another setting. No, thanks!

Then there's the question of how to represent hierarchical bury modes in the GUI. I think 'slider', but we don't have one yet and it would probably fit poorly with the current form layout.
I'm not a fan of GUI elements changing by themselves. The design should prevent invalid configurations in the first place, which is why I'd prefer a dropdown menu.
As for extensibility, I don't think we'd ever want to couple bury modes and sort order. It might make sense from a user perspective that cards shown first should bury later siblings regardless of card state, but the sort order doesn't make a difference for the gathering code. This coupling would only increase code and GUI complexity.
I think if we ever wanted to go back to support more bury configurations, we could just return to the status quo. 4 out of the 2^3 bury modes are 'hierarchical'. If we wanted to support one more mode, we'd need iterative gathering, which could then be easily used to enable all configurations.

**Comment by kevinlawler** [comment]
Still getting this error, crash

````
Anki 25.09.2 (3890e12c) 
Python 3.13.5 Qt 6.9.1 PyQt 6.9.1
Platform: macOS-26.2-arm64-arm-64bit-Mach-O

Traceback (most recent call last):
  File "/Users/user/Library/Application Support/AnkiProgramFiles/.venv/lib/python3.13/site-packages/aqt/taskman.py", line 152, in raise_exception
    raise exception
  File "/Users/user/Library/Application Support/AnkiProgramFiles/.venv/lib/python3.13/site-packages/aqt/taskman.py", line 148, in _on_closures_pending
    closure()
    ~~~~~~~^^
  File "/Users/user/Library/Application Support/AnkiProgramFiles/.venv/lib/python3.13/site-packages/aqt/taskman.py", line 88, in <lambda>
    lambda future: self.run_on_main(lambda: on_done(future))
                                            ~~~~~~~^^^^^^^^
  File "/Users/user/Library/Application Support/AnkiProgramFiles/.venv/lib/python3.13/site-packages/aqt/taskman.py", line 111, in wrapped_done
    on_done(fut)
    ~~~~~~~^^^^^
  File "/Users/user/Library/Application Support/AnkiProgramFiles/.venv/lib/python3.13/site-packages/aqt/operations/__init__.py", line 128, in wrapped_done
    self._success(result)
    ~~~~~~~~~~~~~^^^^^^^^
  File "/Users/user/Library/Application Support/AnkiProgramFiles/.venv/lib/python3.13/site-packages/aqt/reviewer.py", line 558, in after_answer
    self._after_answering(ease)
    ~~~~~~~~~~~~~~~~~~~~~^^^^^^
  File "/Users/user/Library/Application Support/AnkiProgramFiles/.venv/lib/python3.13/site-packages/aqt/reviewer.py", line 571, in _after_answering
    self.nextCard()
    ~~~~~~~~~~~~~^^
  File "/Users/user/Library/Application Support/AnkiProgramFiles/.venv/lib/python3.13/site-packages/aqt/reviewer.py", line 251, in nextCard
    self._get_next_v3_card()
    ~~~~~~~~~~~~~~~~~~~~~~^^
  File "/Users/user/Library/Application Support/AnkiProgramFiles/.venv/lib/python3.13/site-packages/aqt/reviewer.py", line 267, in _get_next_v3_card
    output = self.mw.col.sched.get_queued_cards()
  File "/Users/user/Library/Application Support/AnkiProgramFiles/.venv/lib/python3.13/site-packages/anki/scheduler/v3.py", line 55, in get_queued_cards
    return self.col._backend.get_queued_cards(
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
        fetch_limit=fetch_limit, intraday_learning_only=intraday_learning_only
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/Users/user/Library/Application Support/AnkiProgramFiles/.venv/lib/python3.13/site-packages/anki/_backend_generated.py", line 819, in get_queued_cards
    raw_bytes = self._run_command(13, 3, message.SerializeToString())
  File "/Users/user/Library/Application Support/AnkiProgramFiles/.venv/lib/python3.13/site-packages/anki/_backend.py", line 171, in _run_command
    raise backend_exception_to_pylib(err)
anki.errors.InvalidInput: bug: card modified without updating queue: id:1694885398176 card:1766931881 entry:1766767889

**Comment by abdnh** [comment]
@kevinlawler Can you reproduce it without add-ons? I've not seen recent reports of it.

**Comment by kevinlawler** [comment]
I have no add-ons installed. Confirmed via visual inspection of the `Tools -> Add-ons` window.

This error happens once a day for me with an order 10,000 size deck.

**Comment by kevinlawler** [comment]
I suspect the way to solve this longstanding issue is to create a debug validator method that verifies all the assumptions you have about the card sequence, linear if necessary, and run it as necessary in response to actions, even per card.

I did a `Tools -> Check Database` so we'll see if that solves it.

**Comment by kevinlawler** [comment]
`Check Database` did not fix it

````

Anki 25.09.2 (3890e12c) 
Python 3.13.5 Qt 6.9.1 PyQt 6.9.1
Platform: macOS-26.2-arm64-arm-64bit-Mach-O

Traceback (most recent call last):
  File "/Users/user/Library/Application Support/AnkiProgramFiles/.venv/lib/python3.13/site-packages/aqt/taskman.py", line 152, in raise_exception
    raise exception
  File "/Users/user/Library/Application Support/AnkiProgramFiles/.venv/lib/python3.13/site-packages/aqt/taskman.py", line 148, in _on_closures_pending
    closure()
    ~~~~~~~^^
  File "/Users/user/Library/Application Support/AnkiProgramFiles/.venv/lib/python3.13/site-packages/aqt/taskman.py", line 88, in <lambda>
    lambda future: self.run_on_main(lambda: on_done(future))
                                            ~~~~~~~^^^^^^^^
  File "/Users/user/Library/Application Support/AnkiProgramFiles/.venv/lib/python3.13/site-packages/aqt/taskman.py", line 111, in wrapped_done
    on_done(fut)
    ~~~~~~~^^^^^
  File "/Users/user/Library/Application Support/AnkiProgramFiles/.venv/lib/python3.13/site-packages/aqt/operations/__init__.py", line 128, in wrapped_done
    self._success(result)
    ~~~~~~~~~~~~~^^^^^^^^
  File "/Users/user/Library/Application Support/AnkiProgramFiles/.venv/lib/python3.13/site-packages/aqt/reviewer.py", line 558, in after_answer
    self._after_answering(ease)
    ~~~~~~~~~~~~~~~~~~~~~^^^^^^
  File "/Users/user/Library/Application Support/AnkiProgramFiles/.venv/lib/python3.13/site-packages/aqt/reviewer.py", line 571, in _after_answering
    self.nextCard()
    ~~~~~~~~~~~~~^^
  File "/Users/user/Library/Application Support/AnkiProgramFiles/.venv/lib/python3.13/site-packages/aqt/reviewer.py", line 251, in nextCard
    self._get_next_v3_card()
    ~~~~~~~~~~~~~~~~~~~~~~^^
  File "/Users/user/Library/Application Support/AnkiProgramFiles/.venv/lib/python3.13/site-packages/aqt/reviewer.py", line 267, in _get_next_v3_card
    output = self.mw.col.sched.get_queued_cards()
  File "/Users/user/Library/Application Support/AnkiProgramFiles/.venv/lib/python3.13/site-packages/anki/scheduler/v3.py", line 55, in get_queued_cards
    return self.col._backend.get_queued_cards(
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
        fetch_limit=fetch_limit, intraday_learning_only=intraday_learning_only
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/Users/user/Library/Application Support/AnkiProgramFiles/.venv/lib/python3.13/site-packages/anki/_backend_generated.py", line 819, in get_queued_cards
    raw_bytes = self._run_command(13, 3, message.SerializeToString())
  File "/Users/user/Library/Application Support/AnkiProgramFiles/.venv/lib/python3.13/site-packages/anki/_backend.py", line 171, in _run_command
    raise backend_exception_to_pylib(err)
anki.errors.InvalidInput: bug: card modified without updating queue: id:1694885397697 card:1768230883 entry:1768145083
````

**Comment by abdnh** [comment]
Can you share a sample deck/collection with scheduling data?

**Comment by kevinlawler** [comment]
(Some offline discussion.)

I use Undo (cmd+z), possibly there is a bad interaction there.
